source/Coop/Mod/CoopServer.cs

```diff
@@ -154,18 +179,57 @@ public override string ToString()
 
         private void OnClientConnected(ConnectionServer connection)
         {
+            CoopServerSM coopServerSM = new CoopServerSM();
+            m_CoopServerSMs.Add(connection, coopServerSM);
+
+            #region State Machine Callbacks
+            coopServerSM.SendingWorldDataState.OnEntryFrom(coopServerSM.SendWorldDataTrigger, SendInitialWorldData);
+            #endregion
+
             SyncedObjectStore.AddConnection(connection);
             connection.OnClientJoined += Persistence.ClientJoined;
             connection.OnDisconnected += Persistence.Disconnected;
-            connection.OnServerSendingWorldData += m_GameEnvironmentServer.LockTimeControlStopped;
-            connection.OnServerSendedWorldData += m_GameEnvironmentServer.UnlockTimeControl;
+            OnServerSendingWorldData += m_GameEnvironmentServer.LockTimeControlStopped;
+            OnServerSentWorldData += m_GameEnvironmentServer.UnlockTimeControl;
+
+            connection.Dispatcher.RegisterPacketHandler(ReceiveClientRequestWorldData);
+            connection.Dispatcher.RegisterPacketHandler(ReceiveClientDeclineWorldData);
+
+            connection.Dispatcher.RegisterStateMachine(connection, coopServerSM);
         }
 
         private void OnClientDisconnected(ConnectionServer connection, EDisconnectReason eReason)
         {
+            m_CoopServerSMs.Remove(connection);
+
             connection.OnClientJoined -= Persistence.ClientJoined;
             connection.OnDisconnected -= Persistence.Disconnected;
             SyncedObjectStore?.RemoveConnection(connection);
         }
+
+        [GameServerPacketHandler(ECoopServerState.Preparing, EPacket.Client_RequestWorldData)]
+        private void ReceiveClientRequestWorldData(ConnectionBase connection, Packet packet)
+        {
+            Client_RequestWorldData info =
+                Client_RequestWorldData.Deserialize(new ByteReader(packet.Payload));
+            Logger.Info("Client requested world data.");
+            ((CoopServerSM)m_CoopServerSMs[(ConnectionServer)connection]).StateMachine.Fire(
```

Comment:
Couldn't you make `m_CoopServerSMs` a `Dictionary<ConnectionServer, CoopServerSM>`? That would eliminate all those casts. There's only those specific types in there anyways.